### PR TITLE
bash v5.2, readline v8.2, patches

### PR DESCRIPTION
**Pull request by adamjstewart**
Successfully builds on macOS 12.6.1 (arm64) with Apple Clang 14.0.0.

This PR includes the following changes:

- [x] Add bash 5.2
- [x] Add readline 8.2
- [x] Add patch releases for readline

It turns out that readline has the same release style as bash. There are only minor releases, no patch releases. Instead, a series of patches are shipped alongside the source code, and applying these patches creates new patch releases.

### CODE REVIEW

File: var/spack/repos/builtin/packages/bash/package.py [hunk 1]
```diff
@@ -16,16 +16,31 @@ class Bash(AutotoolsPackage, GNUMirrorPackage):
...
     maintainers = ["adamjstewart"]
 
+    version("5.2", sha256="a139c166df7ff4471c5e0733051642ee5556c1cc8a4a78f145583c5c81ab32fb")
     version("5.1", sha256="cc012bc860406dcf42f64431bcd3d2fa7560c02915a601aba9cd597a39329baa")
     version("5.0", sha256="b4a80f2ac66170b2913efbfb9f2594f1f76c7b1afd11f799e22035d63077fb4d")
     version("4.4", sha256="d86b3392c1202e8ff5a423b302e6284db7f8f435ea9f39b5b1b20fd3ac36dfcb")
     version("4.3", sha256="afc687a28e0e24dc21b988fa159ff9dbcf6b7caa92ade8645cc6d5605cd024d4")
 
     depends_on("ncurses")
+    depends_on("readline@8.2:", when="@5.2:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/bash/package.py:26]: CHANGELOG claims that readline 8.1+ is required in bash 5.2+, but building against 8.1 results in the following error:
```
Undefined symbols for architecture arm64:
  "_rl_set_timeout", referenced from:
      _reset_timeout in libbuiltins.a(read.o)
      _set_readline_timeout in libbuiltins.a(read.o)
  "_rl_trim_arg_from_keyseq", referenced from:
      _bash_execute_unix_command in bashline.o
ld: symbol(s) not found for architecture arm64
```
Building against 8.2 solves the issue.

File: var/spack/repos/builtin/packages/bash/package.py [hunk 2]
```diff
@@ -16,16 +16,31 @@ class Bash(AutotoolsPackage, GNUMirrorPackage):
...
     depends_on("readline@5.0:")
     depends_on("iconv")
+    depends_on("gettext")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/bash/package.py:29]: Previous releases built without this, but the latest release fails:
```
/var/folders/j1/68dlgpr91vlgs26vty2c8xk80000gn/T/ajstewart/spack-stage/spack-stage-bash-5.2-y2ki4v2b74dswufi3cdihigzbsnwoedh/spack-src/include/gettext.h:27:11: fatal error: 'libintl.h' file not found
# include <libintl.h>
          ^~~~~~~~~~~
```
Adding this dep solves the issue in 5.2. In 5.1 and older, bash does link to libintl if found, so I think we should keep the dependency for all versions.

File: var/spack/repos/builtin/packages/readline/package.py [hunk 3]
```diff
@@ -18,16 +18,48 @@ class Readline(AutotoolsPackage, GNUMirrorPackage):
     # URL must remain http:// so Spack can bootstrap curl
     gnu_mirror_path = "readline/readline-8.0.tar.gz"
 
-    version("8.1.2", sha256="7589a2381a8419e68654a47623ce7dfcb756815c8fee726b98f90bf668af7bc6")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/readline/package.py:21]: 8.1.2 == 8.1 + 001 and 002 patches.

File: var/spack/repos/builtin/packages/readline/package.py [hunk 4]
```diff
@@ -18,16 +18,48 @@ class Readline(AutotoolsPackage, GNUMirrorPackage):
...
+    version("8.2", sha256="3feb7171f16a84ee82ca18a36d7b9be109a52c04f492a053331d7d1095007c35")
     version("8.1", sha256="f8ceb4ee131e3232226a17f51b164afc46cd0b9e6cef344be87c65962cb82b02")
     version("8.0", sha256="e339f51971478d369f8a053a330a190781acb9864cf4c541060f12078948e461")
     version("7.0", sha256="750d437185286f40a369e1e4f4764eda932b9459b5ec9a731628393dd3d32334")
     version("6.3", sha256="56ba6071b9462f980c5a72ab0023893b65ba6debb4eeb475d7a563dc65cafd43")
 
     depends_on("ncurses")
-    # from url=https://www.linuxfromscratch.org/patches/downloads/readline/readline-6.3-upstream_fixes-1.patch
-    # this fixes a bug that could lead to seg faults in ipython
-    patch("readline-6.3-upstream_fixes-1.patch", when="@6.3")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/readline/package.py:30]: This patch is included in one of the official patches below, so our copy is no longer needed

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by alalazo** [comment]
@spackbot run pipeline

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Review by tldahlgren — APPROVED** [review]
Confirmed (version) sha256.

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by adamjstewart** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/readline/package.py [hunk 5]
```diff
@@ -18,16 +18,48 @@ class Readline(AutotoolsPackage, GNUMirrorPackage):
...
+
+    patches = [
+        ("8.2", "001", "bbf97f1ec40a929edab5aa81998c1e2ef435436c597754916e6a5868f273aff7"),
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/readline/package.py:30]: This is failing for me when I try to build some python packages from `develop`:

```
==> Error: Couldn't find patch for package builtin.readline with sha256: bbf97f1ec40a929edab5aa81998c1e2ef435436c597754916e6a5868f273aff7
```

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/readline/package.py:30]: I just tried it again and it works for me. Maybe try `spack clean -m`?

**Pull request merged**